feat: add `subscribeAndSpyOn`

katharinakoal opened the pull request:
A first proposal to resolve #7 .
As opposed to my [first approach](https://github.com/hirezio/observer-spy/issues/7#issue-638638278), where I tried to hide subscriptions altogether, I thought it would be more consistent with sticking to subscription terminology. The implementation of this turned out to be way more straight forward. I also wanted to discuss this first before updating the docs.
Let me know what you think about it!

Commit:
test: add tests for `subscribeAndSpyOn` function

Commit:
feat: add `subscribeAndSpyOn` factory function

shairez commented:
Thanks so much @katharinakoal ! 

I liked the `unsubscribe` consistency, good idea! (yeah, much simpler impl it seems)

LGTM 👏

katharinakoal commented:
hey @shairez, thank you for your quick reply! I'll add some usage notes to readme then.

Commit:
docs: add usage notes for factory function

Commit:
refactor: resolve minor code styling issues

shairez committed:
test: remmove redundant test

the first test's action is covered by the second test already

shairez committed:
refactor: export `ObserverSpyWithSubscription`

just in case someone will need to type (TypeScript) its spy with the correct type

shairez committed:
docs(readme): reorder `subscribeAndSpyOn` example code

shairez committed:
Merge branch 'master' into pr/katharinakoal/8

shairez commented:
Thanks @katharinakoal !

I've done a few minor changes (removed a test that I thought was redundant and rearranged the README)

Let me know what you think 

I'm going to merge this and worst case we'll add back that test :)

Thanks again, you are now an official contributor! 🎉👏🙏

shairez commented:
@all-contributors please add @katharinakoal for their code tests documentation ideas and bug reports

shairez commented:
:tada: This PR is included in version 1.3.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@hirez_io/observer-spy/v/1.3.0)
- [GitHub release](https://github.com/hirezio/observer-spy/releases/tag/v1.3.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

katharinakoal commented:
Thanks for polishing up, @shairez !
I'm totally fine with removing the first test. I wanted to check there if the `ObserverSpyWithSubscription` gives you at least the same API as the regular `ObserverSpy`. This is indeed a bit redundant or somehow non-critical since this gets statically checked by TypeScript 🙏.
So it was my pleasure. I go spying now...

shairez commented:
Awesome! Thanks and happy spying! 😎😀
